plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java

```diff
@@ -311,18 +312,14 @@ private static String buildSplitPath(String tableLocation, AddFileEntry addActio
         // paths are relative to the table location and are RFC 2396 URIs
         // https://github.com/delta-io/delta/blob/master/PROTOCOL.md#add-file-and-remove-file
         URI uri = URI.create(addAction.getPath());
-        String path = uri.getPath();
 
-        // org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem encodes the path as URL when opening files
-        // https://issues.apache.org/jira/browse/HADOOP-18580
-        if (tableLocation.startsWith("abfs://") || tableLocation.startsWith("abfss://")) {
-            // Replace '+' with '%2B' beforehand. Otherwise, the character becomes a space ' ' by URL decode.
-            path = URLDecoder.decode(path.replace("+", "%2B"), UTF_8);
+        // decoding is implicit within URI class, done lazily on get methods
```

Comment:
This breaks the existing `TestDeltaLakeAdlsConnectorSmokeTest.testPathUriDecoding` test.

Reply:
@ebyhr 

These are the following tests i carried out
```
-------------------------------- spark:delta-oss:2.3.0 -----------------------

----> same spark queries are being used which is generated by smoke-test

scala> spark.sql("create table dev_30.uri_test (part string, y long) USING delta PARTITIONED BY (part)");

spark.sql("INSERT INTO dev_30.uri_test VALUES ('a=equal', 1), ('a:colon', 2), ('a+plus', 3), ('a space', 4), ('a%percent', 5)").show()

scala> spark.sql("select * from dev_30.uri_test").show(false)
+---------+---+
|part     |y  |
+---------+---+
|a+plus   |3  |
|a space  |4  |
|a%percent|5  |
|a=equal  |1  |
|a:colon  |2  |
+---------+---+

sshuser@shell:~$ hdfs dfs -ls abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/
Found 6 items
drwxr-xr-x   - sshuser sshuser          0 2023-04-28 10:53 abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/_delta_log
drwxr-xr-x   - sshuser sshuser          0 2023-04-28 10:54 abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a space
drwxr-xr-x   - sshuser sshuser          0 2023-04-28 10:54 abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a%25percent
drwxr-xr-x   - sshuser sshuser          0 2023-04-28 10:54 abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a%3Acolon
drwxr-xr-x   - sshuser sshuser          0 2023-04-28 10:54 abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a%3Dequal
drwxr-xr-x   - sshuser sshuser          0 2023-04-28 10:54 abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a+plus

sshuser@shell:~$  hdfs dfs -cat abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/_delta_log/00000000000000000001.json
{"commitInfo":{"timestamp":1682679270014,"operation":"WRITE","operationParameters":{"mode":"Append","partitionBy":"[]"},"readVersion":0,"isolationLevel":"Serializable","isBlindAppend":true,"operationMetrics":{"numFiles":"5","numOutputRows":"5","numOutputBytes":"2375"},"engineInfo":"Apache-Spark/3.3.2 Delta-Lake/2.3.0","txnId":"f68a01da-f7bd-4fa7-8ec8-65b538807c56"}}
{"add":{"path":"part=a%253Dequal/part-00000-66dbb780-45f0-4bdd-9c36-31af0dfb198b.c000.snappy.parquet","partitionValues":{"part":"a=equal"},"size":475,"modificationTime":1682679269000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"y\":1},\"maxValues\":{\"y\":1},\"nullCount\":{\"y\":0}}"}}
{"add":{"path":"part=a%253Acolon/part-00001-86c64e4d-8ecf-445c-a8d6-eb1d8522cd14.c000.snappy.parquet","partitionValues":{"part":"a:colon"},"size":475,"modificationTime":1682679269000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"y\":2},\"maxValues\":{\"y\":2},\"nullCount\":{\"y\":0}}"}}
{"add":{"path":"part=a+plus/part-00002-90e7a0b5-5fd2-42db-b16b-29b0ec63cef2.c000.snappy.parquet","partitionValues":{"part":"a+plus"},"size":475,"modificationTime":1682679269000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"y\":3},\"maxValues\":{\"y\":3},\"nullCount\":{\"y\":0}}"}}
{"add":{"path":"part=a%20space/part-00003-8ec446b3-9556-44c9-9c57-e6bae4b7b7bd.c000.snappy.parquet","partitionValues":{"part":"a space"},"size":475,"modificationTime":1682679269000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"y\":4},\"maxValues\":{\"y\":4},\"nullCount\":{\"y\":0}}"}}
{"add":{"path":"part=a%2525percent/part-00004-fd2f3e1b-5cc6-4d21-a9ea-2aec037e2a28.c000.snappy.parquet","partitionValues":{"part":"a%percent"},"size":475,"modificationTime":1682679269000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"y\":5},\"maxValues\":{\"y\":5},\"nullCount\":{\"y\":0}}"}}

```

```
---------------- reading from trino local build -------

trino> select *, "$path" from delta_prod.dev_30.uri_test;
   part    | y |                                                                                      $path                                                                                       
-----------+---+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 a space   | 4 | abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a space/part-00003-8ec446b3-9556-44c9-9c57-e6bae4b7b7bd.c000.snappy.parquet     
 a:colon   | 2 | abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a%3Acolon/part-00001-86c64e4d-8ecf-445c-a8d6-eb1d8522cd14.c000.snappy.parquet   
 a+plus    | 3 | abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a+plus/part-00002-90e7a0b5-5fd2-42db-b16b-29b0ec63cef2.c000.snappy.parquet      
 a%percent | 5 | abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a%25percent/part-00004-fd2f3e1b-5cc6-4d21-a9ea-2aec037e2a28.c000.snappy.parquet 
 a=equal   | 1 | abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/uri_test/part=a%3Dequal/part-00000-66dbb780-45f0-4bdd-9c36-31af0dfb198b.c000.snappy.parquet
```

The issue here is with the data generated by smoke-test inserts, this is how layout and JSON looks like

```
sshuser@shell:~$ hdfs dfs -ls abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/
Found 7 items
-rw-r-----   1 sshuser sshuser        489 2023-04-28 10:29 abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/README.md
drwxr-x---   - sshuser sshuser          0 2023-04-28 10:29 abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/_delta_log
drwxr-x---   - sshuser sshuser          0 2023-04-28 10:29 abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/part=a space
drwxr-x---   - sshuser sshuser          0 2023-04-28 10:29 abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/part=a%percent
drwxr-x---   - sshuser sshuser          0 2023-04-28 10:29 abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/part=a+plus
drwxr-x---   - sshuser sshuser          0 2023-04-28 10:29 abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/part=a:colon
drwxr-x---   - sshuser sshuser          0 2023-04-28 10:29 abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/part=a=equal

sshuser@shell:~$ hdfs dfs -cat abfs://container@storage.dfs.core.windows.net/test-delta-lake-integration-smoke-test-mtv1h3qaoc/test_uri_table_ojgtzal57z/_delta_log/00000000000000000005.json
{"add":{"path":"part=a%2525percent/part-00000-f6243f45-fea9-4d0e-8744-ef23cfe2b99c.c000.snappy.parquet","partitionValues":{"part":"a%percent"},"size":475,"modificationTime":1671166390000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"y\":5},\"maxValues\":{\"y\":5},\"nullCount\":{\"y\":0}}"}}
{"commitInfo":{"timestamp":1671166390366,"operation":"WRITE","operationParameters":{"mode":"Append","partitionBy":"[]"},"readVersion":4,"isolationLevel":"Serializable","isBlindAppend":true,"operationMetrics":{"numFiles":"1","numOutputRows":"1","numOutputBytes":"475"},"engineInfo":"Apache-Spark/3.3.0 Delta-Lake/2.1.0","txnId":"32d58bfb-1a83-4721-b5ed-b6a2d807c51a"}}
```

^^ As you can see from the first example insert,  the Json's path and the actual partition path both are properly encoded, but in the second example (smoke-test's) the JSON's path is proper whereas the partition data path should be `part=a%25percent` but instead it's `part=a%percent`, thus the issue.

Reply:
@alexjo2144 could you please chime in here?

Reply:
I've followed the guideline from https://github.com/trinodb/trino/blob/master/plugin/trino-delta-lake/src/test/resources/databricks/uri/README.md?plain=1

to create the table on Databricks on top of AWS S3

![image](https://user-images.githubusercontent.com/57260417/236397312-3d1970db-8134-4351-8c7e-dd92cc53c95f.png)


Looking at the outcome of the `INSERT` statement 

```
INSERT INTO findinpathuritest VALUES ('a%percent', 5);
```
I see the following relevant transaction log content (same as in the resources used to register the table):

```
{
  "add": {
    "path": "part=a%2525percent/part-00000-9581ae38-973e-4ce6-989e-8125b474319a.c000.snappy.parquet",
    "partitionValues": {
      "part": "a%percent"
    },
    "size": 597,
    "modificationTime": 1683270576000,
    "dataChange": true,
    "stats": "{\"numRecords\":1,\"minValues\":{\"y\":5},\"maxValues\":{\"y\":5},\"nullCount\":{\"y\":0}}",
    "tags": {
      "INSERTION_TIME": "1683270576000000",
      "MIN_INSERTION_TIME": "1683270576000000",
      "MAX_INSERTION_TIME": "1683270576000000",
      "OPTIMIZE_TARGET_SIZE": "268435456"
    }
  }
```

As we can see on the `path` value `part=a%2525percent/part-00000-9581ae38-973e-4ce6-989e-8125b474319a.c000.snappy.parquet`, the `%` gets a rather curious encoding `%2525`.

Reply:
@vinay-kl We need also to fix tests so encoded paths get created in ABFS. The logic you have removed looks fine in my testing, but I will look more closely tomorrow to verify this is correct.

* Add ABFS specific encoding to smoke test setup

The `TestDeltaLakeAdlsConnectorSmokeTest.testPathUriDecoding` test needs to be modified to correctly test ABFS paths. When uploading to Azure blob storage, URL decode is applied to the filenames ([see here](https://github.com/Azure/azure-sdk-for-java/blob/67ab72ae6d76988cdb7eae0446d227fe0ca77226/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java#L242)). 

The current test uploads paths that look like this on ABFS:
![image](https://github.com/trinodb/trino/assets/4565745/d2fdf7d7-c440-4393-a5f1-5ca89801d9cb)

This is not right - the paths are supposed to be encoded on the filesystem. The tests pass now, because the decode step in DeltaLakeSplitManager that is being removed in this review.

Therefore, to get the tests to pass, there needs to be a [Utility.urlEncode added to TestDeltaLakeAdlsConnectorSmokeTest file upload]  (https://github.com/trinodb/trino/blob/4b17fcf6dc7ff5d540219f59c958d0c350ea34c9/plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAdlsConnectorSmokeTest.java#L131). This has to be the ABFS implementation because it handles `+` and ` ` properly for Azure.
```java
azureContainerClient.getBlobClient(Utility.urlEncode(fileName)).upload(byteSource.openBufferedStream(), byteSource.size());
```

The result will be this:
![image](https://github.com/trinodb/trino/assets/4565745/b4922309-2864-4802-8210-69e4876e02aa)

Reply:
I have verified that if you remove the ABFS custom logic, you can create partitions with `+` and ` `. As long as you fix the tests your change looks good. Please let me know if there's a scenario I am missing or anything I can do to help test this on Azure storage.

Reply:
@ebyhr Do you agree with my suggestion above? Should we should break out the ABFS bug fix from this absolute path feature? i.e. I push a PR to fix testPathUriDecoding and remove the additional URLDecoder.decode?

Reply:
@georgewfisher if just `decode` is removed, the path string may already contain encoded special characters which when passed onto ABFS library gets encoded again, please refer to this https://github.com/trinodb/trino/pull/17038#discussion_r1189390151, i tried to explain with an example.